controlplane/kubeadm/controllers/kubeadm_control_plane_controller_test.go

```diff
@@ -0,0 +1,423 @@
+	kubeadmv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/types/v1beta1"
+	externalv1 "sigs.k8s.io/cluster-api/controllers/external/api/v1alpha3"
+	controlplanev1 "sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/cluster-api/util/kubeconfig"
+	"sigs.k8s.io/cluster-api/util/secret"
+)
+
+func setupScheme() *runtime.Scheme {
+	scheme := runtime.NewScheme()
+	if err := controlplanev1.AddToScheme(scheme); err != nil {
+		panic(err)
+	}
+	if err := clusterv1.AddToScheme(scheme); err != nil {
+		panic(err)
+	}
+	if err := bootstrapv1.AddToScheme(scheme); err != nil {
+		panic(err)
+	}
+	if err := corev1.AddToScheme(scheme); err != nil {
+		panic(err)
+	}
+	if err := externalv1.AddToScheme(scheme); err != nil {
+		panic(err)
+	}
+	return scheme
+}
+
+type fakeClientWrapper struct {
+	fakeClient client.Client
+}
+
+func NewFakeClientWrapperWithScheme(clientScheme *runtime.Scheme, initObjs ...runtime.Object) client.Client {
+	fakeClient := fake.NewFakeClientWithScheme(clientScheme, initObjs...)
+	return &fakeClientWrapper{fakeClient: fakeClient}
+}
+
+func (f fakeClientWrapper) Create(ctx context.Context, obj runtime.Object, opts ...client.CreateOption) error {
+	u, ok := obj.(*unstructured.Unstructured)
+	if ok {
+		if u.GetName() == "" && u.GetGenerateName() != "" {
+			u.SetName(u.GetGenerateName() + util.RandomString(6))
+		}
+		if u.GroupVersionKind().Kind == "GenericMachine" && u.GroupVersionKind().GroupVersion() == externalv1.GroupVersion {
+			gm := &externalv1.GenericMachine{}
+			if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.UnstructuredContent(), gm); err == nil {
+				return f.fakeClient.Create(ctx, gm, opts...)
+			}
+		}
+		return f.fakeClient.Create(ctx, u, opts...)
+	}
+
+	return f.fakeClient.Create(ctx, obj, opts...)
+}
+
+func (f fakeClientWrapper) Update(ctx context.Context, obj runtime.Object, opts ...client.UpdateOption) error {
+	return f.fakeClient.Update(ctx, obj, opts...)
+}
+
+func (f fakeClientWrapper) Delete(ctx context.Context, obj runtime.Object, opts ...client.DeleteOption) error {
+	return f.fakeClient.Delete(ctx, obj, opts...)
+}
+
+func (f fakeClientWrapper) DeleteAllOf(ctx context.Context, obj runtime.Object, opts ...client.DeleteAllOfOption) error {
+	return f.fakeClient.DeleteAllOf(ctx, obj, opts...)
+}
+
+func (f fakeClientWrapper) Patch(ctx context.Context, obj runtime.Object, patch client.Patch, opts ...client.PatchOption) error {
+	return f.fakeClient.Patch(ctx, obj, patch, opts...)
+}
+
+func (f fakeClientWrapper) List(ctx context.Context, list runtime.Object, opts ...client.ListOption) error {
+	return f.fakeClient.List(ctx, list, opts...)
+}
+
+func (f fakeClientWrapper) Get(ctx context.Context, key client.ObjectKey, obj runtime.Object) error {
+	return f.fakeClient.Get(ctx, key, obj)
+}
+
+func (f fakeClientWrapper) Status() client.StatusWriter {
+	return f.fakeClient.Status()
+}
```

Comment:
Ugly workaround for fake client not populating Name when GenerateName is used as well as handling of resources created as unstructured and then subsequently accessed as a typed list.